drivers/spi/spi_ll_stm32.c

```diff
@@ -498,230 +498,54 @@ static int spi_stm32_init(struct device *dev)
-		    &spi_stm32_dev_data_5, &spi_stm32_cfg_5,
-		    POST_KERNEL, CONFIG_SPI_INIT_PRIORITY,
-		    &api_funcs);
-
-#ifdef CONFIG_SPI_STM32_INTERRUPT
-static void spi_stm32_irq_config_func_5(struct device *dev)
-{
-	IRQ_CONNECT(DT_SPI_5_IRQ, DT_SPI_5_IRQ_PRI,
-		    spi_stm32_isr, DEVICE_GET(spi_stm32_5), 0);
-	irq_enable(DT_SPI_5_IRQ);
-}
-#endif
-
-#endif /* CONFIG_SPI_5 */
-
-#ifdef CONFIG_SPI_6
-
-#ifdef CONFIG_SPI_STM32_INTERRUPT
-static void spi_stm32_irq_config_func_6(struct device *port);
-#endif
-
-static const  struct spi_stm32_config spi_stm32_cfg_6 = {
-	.spi = (SPI_TypeDef *) DT_SPI_6_BASE_ADDRESS,
-	.pclken = {
-		.enr = DT_SPI_6_CLOCK_BITS,
-		.bus = DT_SPI_6_CLOCK_BUS
-	},
-#ifdef CONFIG_SPI_STM32_INTERRUPT
-	.irq_config = spi_stm32_irq_config_func_6,
-#endif
-};
-
-static struct spi_stm32_data spi_stm32_dev_data_6 = {
-	SPI_CONTEXT_INIT_LOCK(spi_stm32_dev_data_6, ctx),
-	SPI_CONTEXT_INIT_SYNC(spi_stm32_dev_data_6, ctx),
-};
-
-DEVICE_AND_API_INIT(spi_stm32_6, DT_SPI_6_NAME, &spi_stm32_init,
-		    &spi_stm32_dev_data_6, &spi_stm32_cfg_6,
-		    POST_KERNEL, CONFIG_SPI_INIT_PRIORITY,
-		    &api_funcs);
-
-#ifdef CONFIG_SPI_STM32_INTERRUPT
-static void spi_stm32_irq_config_func_6(struct device *dev)
-{
-	IRQ_CONNECT(DT_SPI_6_IRQ, DT_SPI_6_IRQ_PRI,
-		    spi_stm32_isr, DEVICE_GET(spi_stm32_6), 0);
-	irq_enable(DT_SPI_6_IRQ);
+#define STM32_SPI_IRQ_HANDLER_DECL(id)					\
+	static void spi_stm32_irq_config_func_##id(struct device *dev)
+#define STM32_SPI_IRQ_HANDLER_FUNC(id)					\
+	.irq_config = spi_stm32_irq_config_func_##id,
+#define STM32_SPI_IRQ_HANDLER(id)					\
+static void spi_stm32_irq_config_func_##id(struct device *dev)		\
+{									\
+	IRQ_CONNECT(DT_SPI_##id##_IRQ, DT_SPI_##id##_IRQ_PRI,		\
+		    spi_stm32_isr, DEVICE_GET(spi_stm32_##id), 0);	\
+	irq_enable(DT_SPI_##id##_IRQ);					\
 }
+#else
+#define STM32_SPI_IRQ_HANDLER_DECL(id)
+#define STM32_SPI_IRQ_HANDLER_FUNC(id)
+#define STM32_SPI_IRQ_HANDLER(id)
 #endif
 
-#endif /* CONFIG_SPI_6 */
+#define __STM32_SPI_INIT(id)						\
+STM32_SPI_IRQ_HANDLER_DECL(id);						\
+									\
+static const struct spi_stm32_config spi_stm32_cfg_##id = {		\
+	.spi = (SPI_TypeDef *) DT_SPI_##id##_BASE_ADDRESS,		\
+	.pclken = {							\
+		.enr = DT_SPI_##id##_CLOCK_BITS,			\
+		.bus = DT_SPI_##id##_CLOCK_BUS				\
+	},								\
+	STM32_SPI_IRQ_HANDLER_FUNC(id)					\
+};									\
+									\
+static struct spi_stm32_data spi_stm32_dev_data_##id = {		\
+	SPI_CONTEXT_INIT_LOCK(spi_stm32_dev_data_##id, ctx),		\
+	SPI_CONTEXT_INIT_SYNC(spi_stm32_dev_data_##id, ctx),		\
+};									\
+									\
+DEVICE_AND_API_INIT(spi_stm32_##id, DT_SPI_##id##_NAME, &spi_stm32_init, \
+		    &spi_stm32_dev_data_##id, &spi_stm32_cfg_##id,	\
+		    POST_KERNEL, CONFIG_SPI_INIT_PRIORITY,		\
+		    &api_funcs);					\
+									\
+STM32_SPI_IRQ_HANDLER(id)						\
+
+#define STM32_SPI_INIT(id)						\
+	COND_CODE_1(CONFIG_SPI_##id, (__STM32_SPI_INIT(id)), ())
```

Comment:
While I like the idea, following my experience of users question support, this is likely to confuse users even more, specially because it won't be possible to grep CONFIG_SPI_X anymore.
So please remove this "optimization".

Reply:
I guess that there are many other things that those macros "hide". There are clearly disadvantages of this approach, such as using grep. But when trying to reduce the code by template macros I wanted to do that with everything that is possible, including CONFIG_SPI_X as well. Meaning "convert everything or none". However I'll update according to your request if that will help a lot with reducing confusion.

Reply:
I totally get the approach. Thanks for your understanding.
Fyi, CONFIG_DEVICE_X are supposed to disappear once we complete migration to device tree model (I can't say when this will happen though).